be/src/runtime/sender_queue.cpp

```diff
@@ -0,0 +1,957 @@
+    _is_pipeline_level_shuffle = request.has_is_pipeline_level_shuffle() && request.is_pipeline_level_shuffle();
+
+    if (use_pass_through) {
+        ChunkUniquePtrVector swap_chunks;
+        std::vector<size_t> swap_bytes;
+        _recvr->_pass_through_context.pull_chunks(request.sender_id(), &swap_chunks, &swap_bytes);
+        DCHECK(swap_chunks.size() == swap_bytes.size());
+        size_t bytes = 0;
+        for (size_t i = 0; i < swap_chunks.size(); i++) {
+            // The sending and receiving of chunks from _pass_through_context may out of order, and
+            // considering the following sequences:
+            // 1. add chunk_1 to _pass_through_context and send request_1
+            // 2. add chunk_2 to _pass_through_context and send request_2
+            // 3. receive request_1 and get both chunk_1 and chunk_2
+            // 4. receive request_2 and get nothing
+            // So one receiving may receive two or more chunks, and we need to use the chunk's driver_sequence
+            // but not the request's driver_sequence
+            local_chunk_queue.emplace_back(
+                    ChunkItem::create(swap_bytes[i], swap_chunks[i].second, nullptr, std::move(swap_chunks[i].first)));
+            bytes += swap_bytes[i];
+        }
+        total_chunk_bytes += bytes;
+        COUNTER_UPDATE(_recvr->_bytes_pass_through_counter, total_chunk_bytes);
+    } else {
+        if (_is_pipeline_level_shuffle) {
+            for (auto i = 0; i < request.chunks().size(); i++) {
+                auto& pchunk = request.chunks().Get(i);
+                auto driver_sequence = request.driver_sequences(i);
+                int64_t chunk_bytes = pchunk.data().size();
+                local_chunk_queue.emplace_back(ChunkItem::create(chunk_bytes, driver_sequence, nullptr, pchunk));
+                total_chunk_bytes += chunk_bytes;
+            }
+        } else {
+            // NOTE: in the merge scenario, chunk is obtained through try_get_chunk and its return type is not Status.
+            // there is no chance to handle deserialize error, so the lazy deserialization is not supported now,
+            // we can change related interface's defination to do this later.
+            for (auto i = 0; i < request.chunks().size(); i++) {
+                auto& pchunk = request.chunks().Get(i);
+                int64_t chunk_bytes = pchunk.data().size();
+                ChunkUniquePtr chunk = std::make_unique<vectorized::Chunk>();
+                RETURN_IF_ERROR(_deserialize_chunk(pchunk, chunk.get(), &uncompressed_buffer));
+                local_chunk_queue.emplace_back(ChunkItem::create(chunk_bytes, -1, nullptr, std::move(chunk)));
+                total_chunk_bytes += chunk_bytes;
+            }
+        }
+        COUNTER_UPDATE(_recvr->_bytes_received_counter, total_chunk_bytes);
+    }
+
+    if (_is_cancelled) {
+        return Status::OK();
+    }
+
+    wait_timer.start();
+    {
+        std::lock_guard<Mutex> l(_lock);
```

Comment:
Would this lock section be too large?

Reply:
> Would this lock section be too large?

I think this may be a potential problem, but we have no way to remove this lock at present…after this pr, I tend to try streaming rpc and let it solve the problem of maintaining order…